x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/StepsFactory.java

```diff
@@ -0,0 +1,132 @@
+
+            Phase previousPhase = phaseIterator.previous();
+
+            // add `after` step for phase before next
+            if (phase != null) {
+                // after step should have the name of the previous phase since
+                // the index is still in the
+                // previous phase until the after condition is reached
+                Step.StepKey afterStepKey = new Step.StepKey(previousPhase.getName(), PhaseAfterStep.NAME, PhaseAfterStep.NAME);
+                Step phaseAfterStep = new PhaseAfterStep(nowSupplier, phase.getAfter(), afterStepKey, lastStepKey);
+                steps.add(phaseAfterStep);
+                lastStepKey = phaseAfterStep.getKey();
+            }
+
+            phase = previousPhase;
+            List<LifecycleAction> orderedActions = type.getOrderedActions(phase);
+            ListIterator<LifecycleAction> actionIterator = orderedActions.listIterator(orderedActions.size());
+            // add steps for each action, in reverse
+            while (actionIterator.hasPrevious()) {
+                LifecycleAction action = actionIterator.previous();
+                ActionStepsCreator<?> actionStepsGenerator = actionStepsCreators.get(action.getWriteableName());
+                List<Step> actionSteps = actionStepsGenerator.createSteps(action, client, phase.getName(), lastStepKey);
+                ListIterator<Step> actionStepsIterator = actionSteps.listIterator(actionSteps.size());
+                while (actionStepsIterator.hasPrevious()) {
+                    Step step = actionStepsIterator.previous();
+                    steps.add(step);
+                    lastStepKey = step.getKey();
+                }
+            }
+        }
+
+        if (phase != null) {
+            // The very first after step is in a phase before the hot phase so
+            // call this "new"
+            Step.StepKey afterStepKey = new Step.StepKey("new", PhaseAfterStep.NAME, PhaseAfterStep.NAME);
+            Step phaseAfterStep = new PhaseAfterStep(nowSupplier, phase.getAfter(), afterStepKey, lastStepKey);
+            steps.add(phaseAfterStep);
+            lastStepKey = phaseAfterStep.getKey();
+        }
+
+        // init step so that policy is guaranteed to have
+        steps.add(new InitializePolicyContextStep(InitializePolicyContextStep.KEY, lastStepKey));
+
+        Collections.reverse(steps);
+        LOGGER.debug("STEP COUNT: " + steps.size());
+        for (Step step : steps) {
+            LOGGER.debug(step.getKey() + " -> " + step.getNextStepKey());
+        }
+
+        return steps;
+    }
+
+    @FunctionalInterface
+    public interface ActionStepsCreator<A extends LifecycleAction> {
+
+        List<Step> doCreateSteps(A action, Client client, String phaseName, StepKey nextStepKey);
+
+        @SuppressWarnings("unchecked")
+        default List<Step> createSteps(LifecycleAction action, Client client, String phaseName, StepKey nextStepKey) {
```

Comment:
this needs to be final so we need `ActionStepsCreator` to be an abstract class. We don't want the danger of someone implementing this and changing this method because then we cannot maintain our assumptions that the cast will be safe

Reply:
I found the functionalinterface to be super convenient for less boilerplate, I'll see about returning this to abstract